File: careerpilot-frontend/types/application.ts
export type ApplicationStatus =
  | "applied"
  | "reviewing"
  | "interview"
  | "offered"
  | "rejected";

export interface Application {
  id: number;
  job_id: number;
  resume_id: number;
  user_id: number;
  status: ApplicationStatus | string;
  created_at: string;
}

export interface ApplicationCreatePayload {
  job_id: number;
  resume_id: number;
}

export interface ApplicationUpdatePayload {
  status: string;
}

File: backend/app/models/application.py
from datetime import datetime
from typing import TYPE_CHECKING

from sqlalchemy import DateTime, ForeignKey, String
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.sql import func

from app.core.database import Base

if TYPE_CHECKING:
    from app.models.resume import Resume
    from app.models.user import User
    from app.models.job import Job


class Application(Base):
    __tablename__ = "applications"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)

    status: Mapped[str] = mapped_column(
        String(50),
        default="Applied"
    )

    applied_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now()
    )

    user_id: Mapped[int] = mapped_column(
        ForeignKey("users.id", ondelete="CASCADE")
    )

    job_id: Mapped[int] = mapped_column(
        ForeignKey("job.id", ondelete="CASCADE")
    )

    resume_id: Mapped[int] = mapped_column(
        ForeignKey("resumes.id", ondelete="CASCADE")
    )

    user: Mapped["User"] = relationship(
        back_populates="applications"
    )

    job: Mapped["Job"] = relationship(
        back_populates="applications"
    )

    resume: Mapped["Resume"] = relationship(
        back_populates="applications"
    )